DataFormats/Detectors/CTP/src/ctpRateFetcher.cxx

```diff
@@ -0,0 +1,182 @@
+{
+  std::vector<ctp::CTPClass> ctpcls = mConfig->getCTPClasses();
+  std::vector<int> clslist = mConfig->getTriggerClassList();
+  int classIndex = -1;
+  for (size_t i = 0; i < clslist.size(); i++) {
+    if (ctpcls[i].name.find(className) != std::string::npos) {
+      classIndex = i;
+      break;
+    }
+  }
+  if (classIndex == -1) {
+    LOG(warn) << "Trigger class " << className << " not found in CTPConfiguration";
+    return -1.;
+  }
+  auto rate{mScalers->getRateGivenT(timeStamp * 1.e-3, classIndex, inputType)};
+  return pileUpCorrection(rate.second);
+}
+double ctpRateFetcher::fetchCTPratesClassesNoPuCorr(o2::ccdb::BasicCCDBManager* /*ccdb*/, uint64_t timeStamp, int /*runNumber*/, const std::string& className, int inputType)
+{
+  std::vector<ctp::CTPClass> ctpcls = mConfig->getCTPClasses();
+  std::vector<int> clslist = mConfig->getTriggerClassList();
+  int classIndex = -1;
+  for (size_t i = 0; i < clslist.size(); i++) {
+    if (ctpcls[i].name.find(className) != std::string::npos) {
+      classIndex = i;
+      break;
+    }
+  }
+  if (classIndex == -1) {
+    LOG(warn) << "Trigger class " << className << " not found in CTPConfiguration";
+    return -1.;
+  }
+  auto rate{mScalers->getRateGivenT(timeStamp * 1.e-3, classIndex, inputType)};
+  return rate.second;
+}
+double ctpRateFetcher::fetchCTPratesInputs(o2::ccdb::BasicCCDBManager* /*ccdb*/, uint64_t timeStamp, int /*runNumber*/, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return pileUpCorrection(mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second);
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+double ctpRateFetcher::fetchCTPratesInputsNoPuCorr(o2::ccdb::BasicCCDBManager* /*ccdb*/, uint64_t timeStamp, int /*runNumber*/, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second;
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+
+double ctpRateFetcher::pileUpCorrection(double triggerRate)
+{
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "No filling" << std::endl;
+  }
+  auto bfilling = mLHCIFdata->getBunchFilling();
+  std::vector<int> bcs = bfilling.getFilledBCs();
+  double nbc = bcs.size();
+  double nTriggersPerFilledBC = triggerRate / nbc / constants::lhc::LHCRevFreq;
+  double mu = -std::log(1 - nTriggersPerFilledBC);
+  return mu * nbc * constants::lhc::LHCRevFreq;
+}
+
+void ctpRateFetcher::setupRun(int runNumber, o2::ccdb::BasicCCDBManager* ccdb, uint64_t timeStamp)
+{
+  if (runNumber == mRunNumber) {
+    return;
+  }
+  mRunNumber = runNumber;
+  LOG(debug) << "Setting up CTP scalers for run " << mRunNumber;
+  if (mManualCleanup) {
```

Comment:
For curiosity, in which case you would not want to delete the pointers? Isn't it a source of memory leaks?

Reply:
I dont know - that is Max piece.